File: src/s3.ts
import { S3Client, S3ClientConfig, GetObjectCommand, PutObjectCommand, CreateBucketCommand, DeleteBucketCommand } from '@aws-sdk/client-s3'

const AWS_URL = process.env.AWS_URL
const AWS_ACCESS_KEY_ID = process.env.AWS_ACCESS_KEY_ID
const AWS_SECRET_ACCESS_KEY = process.env.AWS_SECRET_ACCESS_KEY

const s3ClientOptions: S3ClientConfig = {
  region: 'us-east-1',
  endpoint: AWS_URL,
  forcePathStyle: true
}

if (AWS_ACCESS_KEY_ID && AWS_SECRET_ACCESS_KEY) {
  s3ClientOptions.credentials = {
    accessKeyId: AWS_ACCESS_KEY_ID,
    secretAccessKey: AWS_SECRET_ACCESS_KEY
  }
}

let _s3Client: S3Client
function getClient(): S3Client {
  if(!_s3Client) {
    _s3Client = new S3Client(s3ClientOptions)
  }

  return _s3Client
}

type CreateS3BucketInput = {
  bucket: string
}
export async function createS3Bucket(input: CreateS3BucketInput): Promise<any> {
  const { bucket } = input
  const s3Client = getClient()
  const response = s3Client.send(new CreateBucketCommand({ Bucket: bucket }))
  return response
}

type DeleteS3BucketInput = {
  bucket: string
}
export async function deleteS3Bucket(input: DeleteS3BucketInput): Promise<any> {
  const { bucket } = input
  const s3Client = getClient()
  const response = s3Client.send(new DeleteBucketCommand({ Bucket: bucket }))
  return response
}

type GetS3ObjectInput = {
  bucket: string
  key: string
}
export async function getS3Object(input: GetS3ObjectInput): Promise<any> {
  const { bucket, key } = input
  const s3Client = getClient()
  const response = await s3Client.send(
    new GetObjectCommand({
      Bucket: bucket,
      Key: key,
    })
  )
  return response
}

type PutS3ObjectInput = {
  body: Buffer
  bucket: string
  key: string
}
export async function putS3Object(input: PutS3ObjectInput): Promise<any> {
  const { body, bucket, key } = input
  const s3Client = getClient()
  const response = await s3Client.send(
    new PutObjectCommand({
      Body: body,
      Bucket: bucket,
      Key: key,
      ContentType: 'text/csv',
    })
  )
  return response
}

File: src/activities/s3.go
package activities

import (
	"bytes"
	"context"
	"log"
	"os"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"

	"go.temporal.io/sdk/activity"
)

var AWS_URL string = os.Getenv("AWS_URL")
var AWS_ACCESS_KEY_ID string = os.Getenv("AWS_ACCESS_KEY_ID")
var AWS_SECRET_ACCESS_KEY string = os.Getenv("AWS_SECRET_ACCESS_KEY")

// TODO: Update all error handling cases
func getClient(ctx context.Context) (*s3.S3, error) {
	logger := activity.GetLogger(ctx)
	sess, err := session.NewSession(&aws.Config{
		Region:           aws.String("us-east-1"),
		Credentials:      credentials.NewStaticCredentials(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, ""),
		S3ForcePathStyle: aws.Bool(true),
		Endpoint:         aws.String(AWS_URL),
	})
	if err != nil {
		logger.Error("Error from getClient:", err)
		return nil, err
	}

	// Create S3 service client
	client := s3.New(sess)

	return client, nil
}

type CreateS3BucketInput struct {
	Bucket string
}

func CreateS3Bucket(ctx context.Context, input CreateS3BucketInput) error {
	logger := activity.GetLogger(ctx)
	s3Client, err := getClient(ctx)
	if err != nil {
		logger.Error("Error from getClient:", err)
		return err
	}
	_, err = s3Client.CreateBucket(&s3.CreateBucketInput{Bucket: &input.Bucket})
	if err != nil {
		log.Fatal("Error from CreateBucket:", err)
		return err
	}
	return nil
}

type PutS3ObjectInput struct {
	Body   []byte
	Bucket string
	Key    string
}

func putS3Object(ctx context.Context, input PutS3ObjectInput) error {
	logger := activity.GetLogger(ctx)
	s3Client, err := getClient(ctx)
	if err != nil {
		logger.Error("Error from getClient:", err)
		return err
	}
	_, err = s3Client.PutObject(&s3.PutObjectInput{
		Bucket: aws.String(input.Bucket),
		Key:    aws.String(input.Key),
		Body:   bytes.NewReader(input.Body),
	})
	if err != nil {
		log.Fatal("Error from s3Client.PutObject:", err)
		return err
	}
	return nil
}

type GetS3ObjectInput struct {
	Bucket string
	Key    string
}

func GetS3Object(ctx context.Context, input GetS3ObjectInput) ([]byte, error) {
	s3Client, _ := getClient(ctx)
	resp, err := s3Client.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(input.Bucket),
		Key:    aws.String(input.Key),
	})
	if err != nil {
		log.Fatal("Error from s3Client.GetS3Object:", err)
		return nil, err
	}
	defer resp.Body.Close()

	body := new(bytes.Buffer)
	_, err = body.ReadFrom(resp.Body)
	return body.Bytes(), err

}

type DeleteS3BucketInput struct {
	Bucket string
}

func DeleteS3Bucket(ctx context.Context, input DeleteS3BucketInput) (*s3.DeleteBucketOutput, error) {
	s3Client, _ := getClient(ctx)
	_, _ = s3Client.DeleteBucket(&s3.DeleteBucketInput{
		Bucket: aws.String(input.Bucket),
	})

	return nil, nil
}

type DeleteS3ObjectInput struct {
	Bucket string
	Key    string
}

func DeleteS3Object(ctx context.Context, input DeleteS3ObjectInput) (*s3.DeleteObjectOutput, error) {
	s3Client, _ := getClient(ctx)
	_, _ = s3Client.DeleteObject(&s3.DeleteObjectInput{
		Bucket: aws.String(input.Bucket),
		Key:    aws.String(input.Key),
	})

	return nil, nil
}
